File: MoveGen.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;




namespace Chess
{
    class MoveGen
    {
        public static List<int> GetMoves(int[] board, int pos)
        {

            int row = Board.GetRowCol(pos)[0];
            int col = Board.GetRowCol(pos)[1];

            int[] player;
            int[] opponent;
            if (Board.white.Contains(board[pos]))
            {
                player = Board.white;
                opponent = Board.black;
            }
            else if (Board.black.Contains(board[pos]))
            {
                player = Board.black;
                opponent = Board.white;
            }
            else
            {
                return null;
            }

            if (Board.pawn.Contains(board[pos]))
            {
                return PawnMoves(board, pos, row, col, opponent);
            }
            else if (Board.rook.Contains(board[pos]))
            {
                return RookMoves(board, row, col, opponent);
            }
            else if (Board.bishop.Contains(board[pos]))
            {
                return BishopMoves(board, row, col, opponent);
            }
            else if (Board.queen.Contains(board[pos]))
            {
                return QueenMoves(board, row, col, opponent);
            }
            else if (Board.king.Contains(board[pos]))
            {
                return KingMoves(board, row, col, opponent);
            }
            else if (Board.knight.Contains(board[pos]))
            {
                return KnightMoves(board, row, col, opponent);
            }
            else
            {
                return null;
            }
        }


        // TODO: Pawn turn into Queen
        public static List<int> PawnMoves(int[] board, int pos, int row, int col, int[] opponent)
        {
            List<int> moves = new List<int>();

            if (Board.white.Contains(board[pos]))
            {
                if (board[pos + 8] == 0)
                {
                    moves.Add(pos + 8);
                    if (row == 1 && board[pos + 16] == 0)
                    {
                        moves.Add(pos + 16);
                    }
                }
                if (col != 7 && opponent.Contains(board[Board.GetPos(row + 1, col + 1)]))
                {
                    moves.Add(Board.GetPos(row + 1, col + 1));
                }
                if (col != 0 && opponent.Contains(board[Board.GetPos(row + 1, col - 1)]))
                {
                    moves.Add(Board.GetPos(row + 1, col - 1));
                }
            }
            if (Board.black.Contains(board[pos]))
            {
                if (board[pos - 8] == 0)
                {
                    moves.Add(pos - 8);
                    if (row == 6 && board[pos - 16] == 0)
                    {
                        moves.Add(pos - 16);
                    }
                }
                if (col != 7 && opponent.Contains(board[Board.GetPos(row - 1, col + 1)]))
                {
                    moves.Add(Board.GetPos(row - 1, col + 1));
                }
                if (col != 0 && opponent.Contains(board[Board.GetPos(row - 1, col - 1)]))
                {
                    moves.Add(Board.GetPos(row - 1, col - 1));
                }
            }

            return moves;
        }


        public static List<int> RookMoves(int[] board, int row, int col, int[] opponent)
        {
            List<int> moves = new List<int>();

            int up = row - 1;
            int down = row + 1;
            int right = col + 1;
            int left = col - 1;

            while (up >= 0)
            {
                if (board[Board.GetPos(up, col)] != 0)
                {
                    if (opponent.Contains(board[Board.GetPos(up, col)]))
                    {
                        moves.Add(Board.GetPos(up, col));
                    }
                    break;
                }
                moves.Add(Board.GetPos(up, col));
                up -= 1;
            }
            while (down <= 7)
            {
                if (board[Board.GetPos(down, col)] != 0)
                {
                    if (opponent.Contains(board[Board.GetPos(down, col)]))
                    {
                        moves.Add(Board.GetPos(down, col));
                    }
                    break;
                }
                moves.Add(Board.GetPos(down, col));
                down += 1;
            }
            while (right <= 7)
            {
                if (board[Board.GetPos(row, right)] != 0)
                {
                    if (opponent.Contains(board[Board.GetPos(row, right)]))
                    {
                        moves.Add(Board.GetPos(row, right));
                    }
                    break;
                }
                moves.Add(Board.GetPos(row, right));
                right += 1;
            }
            while (left >= 0)
            {
                if (board[Board.GetPos(row, left)] != 0)
                {
                    if (opponent.Contains(board[Board.GetPos(row, left)]))
                    {
                        moves.Add(Board.GetPos(row, left));
                    }
                    break;
                }
                moves.Add(Board.GetPos(row, left));
                left -= 1;
            }

            return moves;

        }

        public static List<int> BishopMoves(int[] board, int row, int col, int[] opponent)
        {
            List<int> moves = new List<int>();

            int up = row - 1;
            int down = row + 1;
            int right = col + 1;
            int left = col - 1;
            bool moveRight = true;
            bool moveLeft = true;

            while (up >= 0)
            {
                if (right <= 7 && moveRight)
                {
                    if (board[Board.GetPos(up, right)] != 0)
                    {
                        if (opponent.Contains(board[Board.GetPos(up, right)]))
                        {
                            moves.Add(Board.GetPos(up, right));
                        }
                        moveRight = false;
                    }
                    if (moveRight)
                    {
                        moves.Add(Board.GetPos(up, right));
                    }
                }
                if (left >= 0 && moveLeft)
                {
                    if (board[Board.GetPos(up, left)] != 0)
                    {
                        if (opponent.Contains(board[Board.GetPos(up, left)]))
                        {
                            moves.Add(Board.GetPos(up, left));
                        }
                        moveLeft = false;
                    }
                    if (moveLeft)
                    {
                        moves.Add(Board.GetPos(up, left));
                    }
                }
                up -= 1;
                right += 1;
                left -= 1;
            }

            right = col + 1;
            left = col - 1;
            moveRight = true;
            moveLeft = true;
            while (down <= 7)
            {
                if (right <=7 && moveRight)
                {
                    if (board[Board.GetPos(down, right)] != 0)
                    {
                        if (opponent.Contains(board[Board.GetPos(down, right)]))
                        {
                            moves.Add(Board.GetPos(down, right));
                        }
                        moveRight = false;
                    }
                    if (moveRight)
                    {
                        moves.Add(Board.GetPos(down, right));
                    }
                }
                if (left >= 0 && moveLeft)
                {
                    if (board[Board.GetPos(down, left)] != 0)
                    {
                        if (opponent.Contains(board[Board.GetPos(down, left)]))
                        {
                            moves.Add(Board.GetPos(down, left));
                        }
                        moveLeft = false;
                    }
                    if (moveLeft)
                    {
                        moves.Add(Board.GetPos(down, left));
                    }
                }
                down += 1;
                right += 1;
                left -= 1;
            }

            return moves;
        }

        public static List<int> QueenMoves(int[] board, int row, int col, int[] opponent)
        {
            return RookMoves(board, row, col, opponent).Concat(BishopMoves(board, row, col, opponent)).ToList();
        }

        public static List<int> KingMoves(int[] board,int row, int col, int[] opponent)
        {
            List<int> moves = new List<int>();

            Tuple<int, int>[] coordinates = new Tuple<int, int>[8] { Tuple.Create(1, -1), Tuple.Create(1, 0), Tuple.Create(1, 1), Tuple.Create(0, -1), Tuple.Create(0, 1), Tuple.Create(-1, -1), Tuple.Create(-1, 0), Tuple.Create(-1, 1) };

            foreach (Tuple<int, int> move in coordinates)
            {
                int y = row + move.Item1;
                int x = col + move.Item2;
                if (y >= 0 && y <= 7 && x >= 0 && x <= 7 && (board[Board.GetPos(y, x)] == 0 || opponent.Contains(board[Board.GetPos(y, x)])))
                {
                    moves.Add(Board.GetPos(y, x));
                }
            }

            return moves;
        }

        public static List<int> KnightMoves(int[] board, int row, int col, int[] opponent)
        {
            List<int> moves = new List<int>();

            Tuple<int, int>[] coordinates = new Tuple<int, int>[8] { Tuple.Create(2, -1), Tuple.Create(2, 1), Tuple.Create(1, -2), Tuple.Create(1, 2), Tuple.Create(-1, -2), Tuple.Create(-1, 2), Tuple.Create(-2, -1), Tuple.Create(-2, 1) };

            foreach (Tuple<int, int> move in coordinates)
            {
                int y = row + move.Item1;
                int x = col + move.Item2;
                if (y >= 0 && y <= 7 && x >= 0 && x <= 7 && (board[Board.GetPos(y, x)] == 0 || opponent.Contains(board[Board.GetPos(y, x)])))
                {
                    moves.Add(Board.GetPos(y, x));
                }
            }

            return moves;
        }

        public static List<int[]> AllMoves(int[] board, int[] player)
        {
            List<int[]> moves = new List<int[]>();

            for (int pos = 0; pos < 64; pos++)
            {
                if (player.Contains(board[pos]))
                {
                    foreach (int newPos in GetMoves(board, pos))
                    {
                        int[] move = new int[2] { pos, newPos };
                        moves.Add(move);
                    }
                }
            }

            return moves;
        }

        public static int[] MakeMoveCopy(int[] board, int pos, int newPos)
        {
            int[] newBoard = board;
            //newBoard = board;
            return MakeMove(newBoard, pos, newPos);
            
        }

        public static int[] MakeMove(int[] board, int pos, int newPos)
        {
            int val = board[pos];
            board[pos] = 0;
            board[newPos] = val;

            return board;
        }






    }


}

File: src/Console Application/MoveGen.py
from board import *


def GetPieceLegalMoves(board, position):

    output = []
    row = GetRowCol(position)[0]
    col = GetRowCol(position)[1]

    up = row + 1
    down = row - 1
    right = col + 1
    left = col - 1

    if board[position] in white:
        player = white
        opposition = black
    elif board[position] in black:
        player = black
        opposition = white
    else:
        return None

    if board[position] in pawn:

        if board[position] in white:
            if row == 7:
                board[position] = 14
                return GetPieceLegalMoves(board, position)

            if board[position + 8] == 0:
                output.append(position + 8)

            if col != 7:
                if board[GetPosition(row + 1, col + 1)] in opposition:
                    output.append(GetPosition(row + 1, col + 1))
            if col != 0:
                if board[GetPosition(row + 1, col - 1)] in opposition:
                    output.append(GetPosition(row + 1, col - 1))

        if board[position] in black:
            if row == 0:
                board[position] = 24
                return GetPieceLegalMoves(board, position)

            if board[position - 8] == 0:
                output.append(position - 8)

            if col != 7:
                if board[GetPosition(row - 1, col + 1)] in opposition:
                    output.append(GetPosition(row - 1, col + 1))
            if col != 0:
                if board[GetPosition(row - 1, col - 1)] in opposition:
                    output.append(GetPosition(row - 1, col - 1))

        return output

    if board[position] in rook:

        while up <= 7:
            if board[GetPosition(up, col)] != 0:
                if board[GetPosition(up, col)] in opposition:
                    output.append(GetPosition(up, col))
                break
            output.append(GetPosition(up, col))
            up += 1
        while down >= 0:
            if board[GetPosition(down, col)] != 0:
                if board[GetPosition(down, col)] in opposition:
                    output.append(GetPosition(down, col))
                break
            output.append(GetPosition(down, col))
            down -= 1
        while right <= 7:
            if board[GetPosition(row, right)] != 0:
                if board[GetPosition(row, right)] in opposition:
                    output.append(GetPosition(row, right))
                break
            output.append(GetPosition(row, right))
            right += 1
        while left >= 0:
            if board[GetPosition(row, left)] != 0:
                if board[GetPosition(row, left)] in opposition:
                    output.append(GetPosition(row, left))
                break
            output.append(GetPosition(row, left))
            left -= 1

        return output

    if board[position] in bishop:

        while up <= 7:
            if right <= 7:
                if board[GetPosition(up, right)] != 0:
                    if board[GetPosition(up, right)] in opposition:
                        output.append(GetPosition(up, right))
                    break
                output.append(GetPosition(up, right))
            up += 1
            right += 1
            left -= 1

        up = row + 1
        right = col + 1
        left = col - 1
        while up <= 7:
            if left >= 0:
                if board[GetPosition(up, left)] != 0:
                    if board[GetPosition(up, left)] in opposition:
                        output.append(GetPosition(up, left))
                    break
                output.append(GetPosition(up, left))
            up += 1
            right += 1
            left -= 1

        right = col + 1
        left = col - 1
        while down >= 0:
            if right <= 7:
                if board[GetPosition(down, right)] != 0:
                    if board[GetPosition(down, right)] in opposition:
                        output.append(GetPosition(down, right))
                    break
                output.append(GetPosition(down, right))
            down -= 1
            right += 1
            left -= 1

        right = col + 1
        left = col - 1
        down = row - 1
        while down >= 0:
            if left >= 0:
                if board[GetPosition(down, left)] != 0:
                    if board[GetPosition(down, left)] in opposition:
                        output.append(GetPosition(down, left))
                    break
                output.append(GetPosition(down, left))
            down -= 1
            right += 1
            left -= 1

        return output

    if board[position] in king:
        for i, j in [(1, -1,), (1, 0), (1, 1), (0, -1), (0, 1), (-1, -1), (-1, 0), (-1, 1)]:
            if 0 <= row+i <= 7 and 0 <= col+j <= 7 and (board[GetPosition(row + i, col + j)] == 0 or board[GetPosition(row + i, col + j)] in opposition):
                output.append(GetPosition(row + i, col + j))

    if board[position] in knight:
        for i, j in [(2, -1), (2, 1), (1, -2), (1, 2), (-1, -2), (-1, 2), (-2, -1), (-2, 1)]:
             if 0 <= row + i <= 7 and 0 <= col + j <= 7 and (board[GetPosition(row + i, col + j)] == 0 or board[GetPosition(row + i, col + j)] in opposition):
                output.append(GetPosition(row + i, col + j))

    if board[position] in queen:

        temp = board[position]

        board[position] = player[2]
        bishop_moves = GetPieceLegalMoves(board, position)
        board[position] = player[3]
        rook_moves = GetPieceLegalMoves(board, position)
        board[position] = temp

        return bishop_moves + rook_moves

    return output


def AllMoves(board, player):

    output = []
    for pos in range(len(board)):
        if board[pos] in player:
            for new_pos in GetPieceLegalMoves(board, pos):
                output.append([pos, new_pos])

    return output


def makeMoveCopy(board, pos, new_pos):
    new_board = list(board)
    return makeMove(new_board, pos, new_pos)


def makeMove(board, pos, new_pos):

    val = board[pos]
    board[pos] = 0
    board[new_pos] = val

    return board












